components/nifosg/nifloader.cpp

```diff
@@ -1319,10 +1327,43 @@ namespace NifOsg
                 return nullptr;
             }
 
-            unsigned char* data = new unsigned char[pixelData->data.size()];
-            memcpy(data, pixelData->data.data(), pixelData->data.size());
+            const std::vector<unsigned char>& pixels = pixelData->data;
+            if (pixelData->fmt != Nif::NiPixelData::NIPXFMT_PAL8)
+            {
+                unsigned char* data = new unsigned char[pixels.size()];
+                std::copy(pixels.begin(), pixels.end(), data);
+
+                image->setImage(width, height, 1, pixelformat, pixelformat, GL_UNSIGNED_BYTE, data, osg::Image::USE_NEW_DELETE);
+            }
+            else
+            {
+                // We're going to convert the indices that pixel data contains
+                // into real colors using the palette.
+                const std::vector<unsigned int>& palette = pixelData->palette->colors;
+
+                std::vector<unsigned char> converted;
+                converted.reserve(pixels.size() * 4);
+                for (size_t i = 0; i < pixels.size(); i++)
+                {
+                    unsigned int color = 0;
+                    if (pixels[i] < palette.size())
+                        color = palette[pixels[i]];
+                    else
+                    {
+                        Log(Debug::Info) << "Color index is not present in the palettized texture palette, ignoring texture";
+                        return nullptr;
+                    }
+                    converted.emplace_back((color >>  0) & 0xFF);
+                    converted.emplace_back((color >>  8) & 0xFF);
+                    converted.emplace_back((color >> 16) & 0xFF);
+                    converted.emplace_back((color >> 24) & 0xFF);
+                }
+                unsigned char* data = new unsigned char[converted.size()];
+                std::copy(converted.begin(), converted.end(), data);
```

Comment:
Why do you need to copy? Just use `data` instead of `converted` variable for transformation destination.

Reply:
OK.